enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The purpose of this enhancement is to outline how the [Operator Lifecycle Manager (OLM)][olm] can let operator users opt-in to the clean up of their operands, or operator workloads, upon the uninstallation of their operator. Specifically when uninstalling an operator, OLM would first delete all CustomResources (CRs) managed by an operator to provide the operator time to clean up all associated resources before deleting the operator itself.
+
+## Motivation
+
+By design when OLM uninstalls an operator it does not remove any of the operator’s owned CustomResourceDefinitions (CRDs) and CRs in order to prevent data loss.
+The operator user or cluster admin is expected to manually delete CRDs, CRs and any other secondary resources created by the operator.
+While some secondary resources that have their [OwnerReference][owner-reference] set to a CR would be garbage collected, others such as off-cluster or cross-namespaced resources with no owner references would remain orphaned.
+
+Without the operator around to process deletion events of CRs, it becomes hard to cleanly remove deployed resources, especially when they exist outside of the cluster.
+
+### Goals
+
+- Clean up includes removal of CRs (operands) and CRDs
+- Operands and CRDs are not automatically removed
+- Upon operator uninstall a user can opt-in to the clean up behavior
+- The operand removal should have a timeout
+- After the operand removal timeout the operator should still be uninstalled
+- Exceeding the operand removal timeout should create an alert event
+
+### Non-Goals
+
+- OLM handles cleanup of anything other than CRs and CRDs
+- Handle custom cleanup logic like filtering/ordering the deletion of CRs 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator user I can instruct OLM to clean up all of my operator's CRs and CRDs when I uninstall my operator.
+
+#### Story 2
+
+As an operator user I want to see an alert event when OLM does not succeed in operand cleanup after a configured timeout.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Specify cleanup via Subscription
+
+To provide a way to opt-in for operand cleanup, the [`Subscription`][subscription] spec can include an optional field `spec.operandCleanup` which lets a user configure the cleanup behaviour.
+
+OLM would initiate cleanup upon the deletion of a `Subscription` instance, based on `spec.operandCleanup`:
+
+##### None (default)
+
+With `spec.operandCleanup.type: none` OLM will not clean up operands. This is also the default behavior if `spec.operandCleanup` is unspecified to make clean up opt-in.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: none
```

Comment:
I can also see this become a list that can have

```yaml
- customresources
- customresourcedefinitions
- operator
```

That would fine tune the clean-up behavior and allow automatic deletions during cluster destruction to opt out of `customresourcedefinitions` to not cause problems with namespaced Operators.

And the default (if nothing is specified) is just `operator`.

Reply:
@dmesser That's interesting but so far I haven't considered operator removal to be part of the cleanup process.
That's part of the operator uninstall process (as it exists today) which always results in operator removal (unless we now introduce an option to block operator removal indefinitely).

So the list of things to cleanup should just be:
```
- customresources
- customresourcedefinitions
```

Otherwise with the full list:
```
- customresources
- customresourcedefinitions
- operator
```
users could leave out `operator` and now you have the possibility of an operator uninstall process (with cleanup) where you delete the Subscription but the operator stays installed.

Not removing CRDs for namespaced operator cleanups is still something OLM can easily infer and handle.
Otherwise in this UX, the onus is on users to make sure they explicitly omit `customresourcedefinitions` for their cleanup to avoid borking other namespaced operators.

If the list of things to cleanup was larger then I could see the benefit of more fine tuning but right now it's maybe more complexity for the default case.
For more fine tuning/control over cleanup I would lean towards revisiting the API to allow deferring to a custom job or the operator altogether.

Reply:
Makes sense, thank you.